File: Exercises/ResponseTimeAnalysis/src/real_time_analysis.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <fstream>
#include <sstream>
#include <cmath>

using namespace std;

struct Task {
    string id;
    int BCET;
    int WCET;
    int period;  
    int deadline; 
    int priority; // Lower number => higher priority for RMS, if you wish
};


void readTasksCSV(const string& filename, vector<Task>& tasks)
{
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Error: Could not open file " << filename << endl;
        return;
    }

    string line;
    getline(file, line); // Skip header

    while (getline(file, line))
    {
        stringstream ss(line);
        Task task;
        string token;

        getline(ss, token, ','); // Task ID
        task.id = token;
        getline(ss, token, ','); // WCET
        task.WCET = stoi(token);
        getline(ss, token, ','); // BCET
        task.BCET = stoi(token);
        getline(ss, token, ','); // Period
        task.period = stoi(token);
        getline(ss, token, ','); // Deadline
        task.deadline = stoi(token);
        getline(ss, token, ','); // Priority
        task.priority = stoi(token);

        tasks.push_back(task);
    }
    file.close();
}

void RTA_test(vector<Task>& tasks)
{
    // Sort tasks by priority (lower number means higher priority)
    sort(tasks.begin(), tasks.end(), [](const Task& a, const Task& b) {
        return a.priority < b.priority;  // Ensure lower priority number = higher priority
    });

    for (size_t task = 0; task < tasks.size(); task++)
    {
        int R = tasks[task].WCET;
        int last_R = -1;
        bool schedulable = true;
        int I = 0; // Reset interference for each iteration

        while (R != last_R)
        {
            last_R = R;

            // Compute interference from higher-priority tasks
            for (size_t j = 0; j < task; j++)
            {
                I += ceil((double)R / tasks[j].period) * tasks[j].WCET;
            }

            R = I + tasks[task].WCET;

            if (R > tasks[task].deadline)
            {
                cout << "Task " << tasks[task].id << " is not schedulable." << endl;
                schedulable = false;
                break;
            }
        }

        if (schedulable)
        {
            cout << "Task " << tasks[task].id << " is schedulable with WCRT = " << R << endl;
        }
    }
}

int main(int argc, char* argv[])
{
    if (argc != 2) {
        cerr << "Usage: " << argv[0] << " <tasks.csv>" << endl;
        return 1;
    }

    string filename = argv[1];
    vector<Task> tasks;
    readTasksCSV(filename, tasks);

    RTA_test(tasks);

    return 0;
}

File: Exercises/ResponseTimeAnalysis/src/real_time_analysis.py
import pandas as pd
import math


def read_csv(file):
    """
    Read tasks from a CSV file and convert to a list of task dictionaries.
    """
    df = pd.read_csv(file)
    req_columns = ['Task', 'BCET', 'WCET', 'Period', 'Deadline', 'Priority']

    if not all(col in df.columns for col in req_columns):
        raise ValueError(f"Missing required column(s): {req_columns}")

    tasks = []
    for _, row in df.iterrows():
        task = {
            "name": row['Task'],
            "bcet": row['BCET'],
            "wcet": row['WCET'],
            "period": row['Period'],
            "deadline": row['Deadline'],
            "priority": row['Priority']
        }
        tasks.append(task)

    return tasks


def response_time_analysis(tasks):
    """
    Perform Response-Time Analysis.
    """
    # Sort tasks by priority (lower number means higher priority)
    sorted_tasks = sorted(tasks, key=lambda x: x['priority'])

    results = []

    for i, task in enumerate(sorted_tasks):
        R = task['wcet']  # Initial response time
        last_R = -1
        schedulable = True

        while R != last_R:
            last_R = R
            interference = 0

            # Calculate interference from higher-priority tasks
            for j in range(i):
                higher_task = sorted_tasks[j]
                interference += math.ceil(R / higher_task['period']) * higher_task['wcet']

            R = task['wcet'] + interference

            # Check if the task is schedulable
            if R > task['deadline']:
                schedulable = False
                break

        # Store results
        result = {
            'Task': task['name'],
            'WCRT': R,
            'Deadline': task['deadline'],
            'Status': 'Schedulable' if schedulable else 'Not Schedulable'
        }
        results.append(result)

    return results


def run_rta(path_file):
    """
    Main function to read tasks and perform response time analysis.
    """
    # Read tasks from CSV
    tasks = read_csv(path_file)
    print(tasks)
    # Perform response time analysis
    analysis_results = response_time_analysis(tasks)

    # Print results
    for result in analysis_results:
        if result['Status'] == 'Schedulable':
            print(f"Task {result['Task']} is schedulable with WCRT = {result['WCRT']:.1f}")
        else:
            print(f"Task {result['Task']} is not schedulable with WCRT.")
    return analysis_results
